compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1032,6 +1032,54 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Swift lets users write `Ty?` to mean `Option<Ty>`. Parse the construct and recover from it.
+    pub(super) fn maybe_recover_from_question_mark(
+        &mut self,
+        ty: P<Ty>,
+        is_as_cast: IsAsCast,
+    ) -> P<Ty> {
+        if let IsAsCast::Yes = is_as_cast {
+            return ty;
+        }
+        if self.token == token::Question {
+            self.bump();
+            self.struct_span_err(self.prev_token.span, "invalid `?` in type")
+                .span_label(self.prev_token.span, "`?` is only allowed on expressions, not types")
+                .multipart_suggestion(
+                    "if you meant to express that the type might not contain a value, use the `Option` wrapper type",
+                    vec![
+                        (ty.span.shrink_to_lo(), "Option<".to_string()),
+                        (self.prev_token.span, ">".to_string()),
+                    ],
+                    Applicability::MachineApplicable,
+                )
+                .emit();
+
+            // Synthesize an `Option<ty>` from `ty?`.
+            let span = ty.span.to(self.prev_token.span);
+            let path = Path {
+                span,
+                segments: vec![
+                    PathSegment::from_ident(Ident::from_str("std")),
```

Comment:
Would this work with `#![no_std]`?

Reply:
Probably just

```suggestion
                    PathSegment::from_ident(Ident::from_str("core")),
```

could resolve this, assuming the code that trims unnecessary path prefixes works with `core` too.

Reply:
I also think this should add a `PathSegment::path_root()` to the front in case someone has a module named `core.`

Reply:
For some reason using `::core::option::Option<` doesn't work (`core` isn't found`), while `::std::option::Option` and `core::option::Option` do.

Reply:
That's _really_ strange...

Reply:
Using a non-global path seems like it could cause some confusing errors though. I know that I've worked on code with an import named `core` in scope.

Reply:
The other option is to return `TyKind:Err` and side-step the issue entirely.